tpm2/seal.go

```diff
@@ -54,204 +58,163 @@ type ProtectKeyParams struct {
 	// owner objects (0x01800000 - 0x01bfffff).
 	PCRPolicyCounterHandle tpm2.Handle
 
-	// AuthKey is the key used to authorize changes to the newly create key,
-	// via the SealedKeyObject.UpdatePCRProtectionPolicy and
-	// secboot.KeyData.SetAuthorizedSnapModels APIs. If set, this should be a
-	// random 32-byte number.
-	// If not set, one is generated automatically.
-	AuthKey secboot.PrimaryKey
-
-	// AuthorizedSnapModels is a list of models initially authorized to access
-	// the data protected by the newly created key. These can be updated later
-	// on by calling secboot.KeyData.SetAuthorizedSnapModels.
-	AuthorizedSnapModels []secboot.SnapModel
+	PrimaryKey secboot.PrimaryKey
 }
 
-// makeKeyDataWithPolicy protects the supplied keys using the supplied keySealer and
-// policy data. This can be called multiple times to protect an arbitary number of
-// keys with an identical policy.
-func makeKeyDataWithPolicy(key secboot.DiskUnlockKey, authKey secboot.PrimaryKey, policy *keyDataPolicyParams, sealer keySealer) (*secboot.KeyData, error) {
-	var symKey [32 + aes.BlockSize]byte
-	if _, err := rand.Read(symKey[:]); err != nil {
-		return nil, xerrors.Errorf("cannot create symmetric key: %w", err)
-	}
+type PassphraseProtectKeyParams struct {
+	ProtectKeyParams
 
-	priv, pub, importSymSeed, err := sealer.CreateSealedObject(symKey[:], policy.Alg, policy.AuthPolicy)
-	if err != nil {
-		return nil, err
-	}
-
-	// Create a new SealedKeyObject.
-	data, err := newKeyData(priv, pub, importSymSeed, policy.PolicyData)
-	if err != nil {
-		return nil, xerrors.Errorf("cannot create key data: %w", err)
-	}
-	skd := &SealedKeyData{sealedKeyDataBase: sealedKeyDataBase{data: data}}
+	KDFOptions *secboot.KDFOptions
+}
 
-	// Create encrypted payload
-	payload := secboot.MarshalKeys(key, authKey)
+type keyDataConstructor func(skd *SealedKeyData, role string, encryptedPayload []byte, kdfAlg crypto.Hash) (*secboot.KeyData, error)
 
-	b, err := aes.NewCipher(symKey[:32])
-	if err != nil {
-		return nil, xerrors.Errorf("cannot create new cipher: %w", err)
-	}
-	stream := cipher.NewCFBEncrypter(b, symKey[32:])
-	stream.XORKeyStream(payload, payload)
+func (fn keyDataConstructor) NewKeyData(skd *SealedKeyData, role string, encryptedPayload []byte, kdfAlg crypto.Hash) (*secboot.KeyData, error) {
```

Comment:
I don't find the indirection of a method on a func type more readable or idiomatic of simply using the func type alone but maybe I'm missing something

Reply:
I've removed the indirections here.